src/main.rs

```diff
@@ -60,33 +154,65 @@ fn main() {
-    match template_file.read_to_string(&mut template_str) {
-        Err(msg) => panic!("Failed to read {}: {}", display, msg),
-        Ok(_) => println!("Successfully read {}", display),
-    }
 
-    let handlebars = Handlebars::new();
+    // Use the information in the code generator code_gen and format it, using handlebars
+    // to a template with all the basic filter information enclosed
+    let c_mode = matches.value_of("compilation_mode").unwrap();
+    let filter_name = matches.value_of("filter_name").unwrap();
+    if c_mode == "sim" {
+        // Because we are making a library, not just one file, we need to copy over an example library.  Then,
+        // we have to write to three files:  Cargo.toml to edit the filter name, lib.rs to edit the filter name,
+        // and <filter_name>.rs for the actual filter implementation
+        
+        // Making new library folder
+        let mut lib_src_folder = filter_name.to_string();
+        lib_src_folder.push_str("/src/");
+        fs::create_dir_all(&lib_src_folder).unwrap();
+        
+        let src = "src";
 
-    let output = handlebars
-        .render_template(&template_str, &code_gen)
-        .expect("handlebar render failed");
+        // Cargo.toml
+        let cargo_file_name : PathBuf = [filter_name, "Cargo.toml"].iter().collect();
+        let cargo_handlebars_name : PathBuf = [FILTER_FOLDER, "Cargo.toml.handlebars"].iter().collect();
+        generate_code_from_filter_name_with_handlebars(filter_name.to_string(), cargo_handlebars_name, cargo_file_name);
 
-    let mut file = File::create(out_file).expect("file create failed.");
-    file.write_all(output.as_bytes()).expect("write failed");
+        // Filter types
+        let mut filename = String::from(filter_name);
+        filename.push_str("_types.rs");
+        let filter_types_file : PathBuf = [filter_name, &filename].iter().collect();
+        let filter_types_handlebars : PathBuf = [FILTER_FOLDER, "src", "filter_types.rs.handlebars"].iter().collect();
+        generate_code_from_codegen_with_handlebars(&code_gen, filter_types_handlebars, filter_types_file);
+
+        // The lib file
+        let lib_file_name : PathBuf = [filter_name, src, "lib.rs"].iter().collect();
+        let lib_file_handlebars : PathBuf = [FILTER_FOLDER, "src", "lib.rs.handlebars"].iter().collect();
+        generate_code_from_filter_name_with_handlebars(filter_name.to_string(), lib_file_handlebars, lib_file_name);
+        
+        // The filter itself
+        let mut filter_file_name :PathBuf = [filter_name, src, filter_name].iter().collect();
+        filter_file_name.set_extension(".rs");
+        let filter_name_handlebars : PathBuf = [FILTER_FOLDER, "src", "filter.rs.handlebars"].iter().collect();
+        generate_code_from_codegen_with_handlebars(&code_gen, filter_name_handlebars, filter_file_name);
+
+        // The graph_utils file
+        let mut graph_utils_new_file_name = filter_name.to_string().clone();
```

Comment:
What is this for? Is it used?

Reply:
This is a file with functions to deal with the graph part of the filter.  So it can create a graph from the path headers, create a graph from the cql query, and then do different functions to compare them.  It is shipped with the filter library because the filter itself uses these as helper functions.  I put them in a separate file, because I think it's more modular and easier to read;  you don't have to deal with graph logic when you're just reading a filter, trying to figure out what it is doing.